File: jotupix_demo_android/jotupix/src/main/java/com/jtkj/jotupix/program/JTextContent.java
package com.jtkj.jotupix.program;

import com.jtkj.jotupix.utils.JByteWriter;

import java.util.ArrayList;
import java.util.List;

public class JTextContent extends JContentBase {

    public int bgColor;
    public int blendType;
    public int showX;
    public int showY;
    public int showWidth;
    public int showHeight;
    public int showMode;
    public int showSpeed;
    public int stayTime;
    public int moveSpace;
    public int textNum;
    public int textAllWide;

    public List<JTextFont> textData;

    public JTextContent() {
        this.contentType = ContentType.TEXT;
        this.textData = new ArrayList<>();
    }

    @Override
    public byte[] get() {
        JByteWriter content=new JByteWriter();

        // Put content type
        content.putByteOne(contentType);

        // Put 5 bytes of zeros
        content.putRepeatByteOne(0, 5);

        // Put background color
        content.putBytesTwo(bgColor);

        // Put blend type
        content.putByteOne((blendType == BlendType.MIX ? 0 : 1));

        // Put coordinates and dimensions
        content.putBytesTwo(showX);
        content.putBytesTwo(showY);
        content.putBytesTwo(showWidth);
        content.putBytesTwo(showHeight);

        // Put display properties
        content.putByteOne(showMode);
        content.putByteOne(showSpeed);
        content.putByteOne(stayTime);

        // Put movement and text properties
        content.putBytesTwo(moveSpace);
        content.putBytesTwo(textNum);
        content.putBytesFour(textAllWide);

        // Put text font data
        for (JTextFont font : textData) {
            content.putByteOne(font.textWidth);
            content.putByteOne(font.textType);
            content.putBytes(font.showData);
        }

        // Insert total size at the beginning (size + 4)
        content.insertBytesFour(0, content.size() + 4);

        return content.toByteArray();
    }
}

File: jotupix_demo_ios/JotupixDemo/Jotupix/Program/JTextContent.m
#import "JTextContent.h"
#import "JByteWriter.h"

@implementation JTextContent

- (instancetype)init {
    if (self = [super init]) {
        self.contentType = JContentTypeText;  
        _textData = [NSMutableArray array];
    }
    return self;
}

- (NSData *)get {

    JByteWriter *content = [[JByteWriter alloc] init];

    // content type
    [content put_u8:(uint8_t)self.contentType];

    // reserved = 5 bytes
    [content put_repeat:0 count:5];

    [content put_u16:self.bgColor];
    [content put_u8:(uint8_t)self.blendType];
    [content put_u16:self.showX];
    [content put_u16:self.showY];
    [content put_u16:self.showWidth];
    [content put_u16:self.showHeight];
    [content put_u8:self.showMode];
    [content put_u8:self.showSpeed];
    [content put_u8:self.stayTime];
    [content put_u16:self.moveSpace];
    [content put_u16:self.textNum];
    [content put_u32:self.textAllWide];

    // Each font block
    for (JTextFont *font in self.textData) {
        [content put_u8:font.textWidth];
        [content put_u8:(uint8_t)font.textType];
        [content put_bytes:font.showData];
    }

    // Prepend total block size (includes size of this u32)
    uint32_t totalSize = (uint32_t)(content.size + 4);

    // Correct usage: insert_u32(pos, value)
    [content insert_u32:0 value:totalSize];

    // Must return NSData, not uint8_t*
    return [NSData dataWithBytes:content.buffer length:content.size];
}

@end
